### PR TITLE
Show time for current map center

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 1]
```diff
@@ -369,4 +372,24 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
...
+{
+    OsmAnd::LatLon centerLocation = OsmAnd::Utilities::convert31ToLatLon([OARootViewController instance].mapPanel.mapViewController.mapView.target31);
+    CLLocation *newCenterLatLon = [[CLLocation alloc] initWithLatitude:centerLocation.latitude longitude:centerLocation.longitude];
+    if (![self isLocationsEqual:_cachedCenterLatLon with:newCenterLatLon])
+        _cachedCenterLatLon = newCenterLatLon;
+}
+
+- (BOOL) isLocationsEqual:(CLLocation *)firstCoordinate with:(CLLocation *)secondCoordinate
+{
+    if (firstCoordinate && secondCoordinate)
```
Comment (tigrim) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:385]: CLLocationCoordinate2DIsValid

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 2]
```diff
@@ -369,4 +372,24 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
     return [dateFormatter stringFromDate:nextDate];
 }
 
+- (void) updateCachedLocation
+{
+    OsmAnd::LatLon centerLocation = OsmAnd::Utilities::convert31ToLatLon([OARootViewController instance].mapPanel.mapViewController.mapView.target31);
```
Comment (tigrim) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:377]: It's more optimal to cache target31, this way you won't have to create a new CLLocation object each time

**Review by tigrim — APPROVED** [review]

**Review by tigrim — COMMENTED** [review]
review

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 3]
```diff
@@ -369,4 +373,21 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
     return [dateFormatter stringFromDate:nextDate];
 }
 
+- (void) updateCachedLocation
+{
+    OsmAnd::PointI currentTarget31 = [OARootViewController instance].mapPanel.mapViewController.mapView.target31;
```
Comment (alex-osm) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:378]: `OAMapViewController` has `getMapLocation`

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 4]
```diff
@@ -369,4 +373,21 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
...
+        _cachedTarget31 = currentTarget31;
+        OsmAnd::LatLon newCenterLocation = OsmAnd::Utilities::convert31ToLatLon(_cachedTarget31);
+        if (![self isLocationsEqual:_cachedCenterLatLon with:newCenterLocation])
+            _cachedCenterLatLon = newCenterLocation;
+    }
+}
+
+- (BOOL) isLocationsEqual:(OsmAnd::LatLon)firstLocation with:(OsmAnd::LatLon)secondLocation
+{
+    return fabs(firstLocation.longitude - secondLocation.longitude) <= 0.0001;
```
Comment (alex-osm) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:390]: Why latitude not taken into consideration?

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 5]
```diff
@@ -369,4 +373,21 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
     return [dateFormatter stringFromDate:nextDate];
 }
 
+- (void) updateCachedLocation
```
Comment (alex-osm) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:376]: It is not clear purpose of `updateCachedLocation`? Why cached location even needed in such implementation?
The `updateCachedLocation` should return `BOOL` and if location not changed - do not recalculate sunrise / sunset time!

File: Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm [hunk 6]
```diff
@@ -369,4 +409,33 @@ - (NSString *) formatNextTime:(NSTimeInterval)nextTime
...
+{
+    CLLocation *newCenterLatLon = [[OARootViewController instance].mapPanel.mapViewController getMapLocation];
```
Comment (tigrim) [inline Sources/Controllers/Map/Widgets/OASunriseSunsetWidget.mm:414]: it seems your can use CLLocationCoordinate2D loc = [OAResourcesUIHelper getMapLocation]; ?

**Review by tigrim — APPROVED** [review]

**Pull request merged**